File: Web/src/main/java/com/itechart/webflux/web/config/jwt/JwtAuthProvider.java
package com.itechart.webflux.web.config.jwt;

import com.itechart.webflux.web.core.model.entity.User;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.authentication.dao.AbstractUserDetailsAuthenticationProvider;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.AuthorityUtils;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class JwtAuthProvider extends AbstractUserDetailsAuthenticationProvider {

    private static final Logger LOGGER = LoggerFactory.getLogger(JwtAuthProvider.class);

    @Autowired
    private JwtVal validator;

    @Override
    protected void additionalAuthenticationChecks(UserDetails userDetails, UsernamePasswordAuthenticationToken usernamePasswordAuthenticationToken) throws AuthenticationException {
        //empty
    }

    @Override
    protected UserDetails retrieveUser(String s, UsernamePasswordAuthenticationToken usernamePasswordAuthenticationToken) throws AuthenticationException {
        JwtAuthToken jwtAuthenticationToken = (JwtAuthToken) usernamePasswordAuthenticationToken;
        String token = jwtAuthenticationToken.getToken();
        if (token == null) {
            return null;
        }
        User jwtUser = validator.validate(token);

        List<GrantedAuthority> grantedAuthorities = AuthorityUtils
                .commaSeparatedStringToAuthorityList(String.valueOf(jwtUser.getRole()));
        return new UserDetail(jwtUser.getUsername(), token, jwtUser.getId(), grantedAuthorities);

    }
}

File: Web/src/main/java/com/itechart/webflux/web/config/jwt/JwtAuthProvider.kt
package com.itechart.webflux.web.config.jwt

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.authentication.dao.AbstractUserDetailsAuthenticationProvider
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.authority.AuthorityUtils
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component

@Component
class JwtAuthProvider : AbstractUserDetailsAuthenticationProvider() {

    @Autowired
    private val validator: JwtVal? = null

    @Throws(AuthenticationException::class)
    override fun additionalAuthenticationChecks(userDetails: UserDetails, usernamePasswordAuthenticationToken: UsernamePasswordAuthenticationToken) {
        //empty
    }

    @Throws(AuthenticationException::class)
    override fun retrieveUser(s: String, usernamePasswordAuthenticationToken: UsernamePasswordAuthenticationToken): UserDetails {
        val jwtAuthenticationToken = usernamePasswordAuthenticationToken as JwtAuthToken
        val token = jwtAuthenticationToken.token
        val jwtUser = token?.let { validator!!.validate(it) }

        val grantedAuthorities = AuthorityUtils
                .commaSeparatedStringToAuthorityList(jwtUser!!.role.toString())
        return UserDetail(jwtUser.username, token, jwtUser.id, grantedAuthorities)

    }

}